21.py

```diff
@@ -0,0 +1,50 @@
+# https://projecteuler.net/problem=21
+
+def is_proper_divisor(divisor, n):
+    if divisor >= n:
+        return False
+    if n % divisor == 0:
+        return True
+    return False
+
+
+def sum_proper_divisors(n):
+    sum = 0
+
+    for x in range(1, n + 1):
+        if is_proper_divisor(x, n):
+            sum += x
+
+    return sum
+
+
+def divisor_sums_in_range(n):
+    number_pairs = []
+
+    for x in range(1, n + 1):
+        number_pairs.append((x, sum_proper_divisors(x)))
+
+    return number_pairs
+
+
+def is_amicable_pair(p1, p2):
+    if p1[0] == p2[1] and p1[1] == p2[0]:
+        return True
+    return False
+
+# would be better if this function removed items from the list as they were checked
+def sum_amicable_pairs(n):
+    amicable_pairs = 0
+
+    divisor_sums = divisor_sums_in_range(n)
+    for x in divisor_sums:
```

Comment:
i think another thing to consider is that these amicable pairs are literal pairs, there's never another buddy along with them.

if d(x) = y and d(y) = x, then d(z) can never be x or y. there's no need to check a n^2 loop here.

you should probably look from 1..n and do something like

```py
potentional_pair = d(x)
if (if d(potentional_pair) == x): # we have a winner
```

your comment about removing the pair once you found it is probably good too, but that doesn't fix the immediate problem. the calculations inside the loop is so minimal (just checking a boolean basically) that the duplicates aren't the issue

idk if this really applies to your method here but it might help

Reply:
that makes a lot of sense, i was trying to figure out how to get rid of the n^2 loop. i will give this a shot

Reply:
that did it! the script returns the right result in < 1s now, whereas before with the n^2 loop it was taking 20 seconds +